JENKINS-45240# Change github pipeline creation to multi-branch pipeline

vivek opened the pull request:
# Description

Removes organization folder based github/ghe pipeline creation with multi-branch one. End to end GitHub pipeline creation supported in this PR. Includes snapshot released dependency of pipeline editor, see https://github.com/jenkinsci/blueocean-pipeline-editor-plugin/pull/53. 

See [JENKINS-45240](https://issues.jenkins-ci.org/browse/JENKINS-45240).

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

i386 commented:
Deleting more code than you added 👍

michaelneale commented:
works nicely @vivek !

I did see an oddness when I went to add the same repo again - it correctly prompted me to provide a non duplicate name, but then after a few seconds that field went grey - perhaps @cliffmeyers knows why? 

![screen shot 2017-08-18 at 3 30 26 pm](https://user-images.githubusercontent.com/14976/29445520-aaddcbf0-842a-11e7-83e7-c0d424f27343.png)

cliffmeyers commented on blueocean-dashboard/src/main/js/creation/github/GithubFlowManager.js:
```diff
@@ -52,40 +50,15 @@ export default class GithubFlowManager extends FlowManager {
-        return this.repositories.filter(repo => (
-            this.existingOrgFolder.pipelineFolderNames.indexOf(repo.name) === -1
-        ));
+        return this.repositories;
```
Perhaps we should just refactor all references to `selectableRepositories` to `repositories` instead?

cliffmeyers commented on blueocean-dashboard/src/main/js/creation/github/GithubFlowManager.js:
```diff
@@ -96,10 +69,6 @@ export default class GithubFlowManager extends FlowManager {
-    savedOrgFolder = null;
-
-    savedPipeline = null;
-
```
I believe I saw a reference to a `pipeline` property in another file... might be good to initialize it here.

cliffmeyers replied:
`pipelineName` too

vivek replied:
ok, will do.

cliffmeyers commented on blueocean-dashboard/src/main/js/creation/github/GithubFlowManager.js:
```diff
@@ -140,7 +117,7 @@ export default class GithubFlowManager extends FlowManager {
     getApiUrl() {
         // backend will default to api.github.com
-        return null;
+        return 'https://api.github.com';
```
Any reason to change it from null? I've noticed that when `apiUrl=https://api.github.com` is sent to creation endpoint, it adds it as an endpoint but this always seems odd to me.

vivek replied:
yeah MBP pipeline creation backend requires it.

cliffmeyers commented on blueocean-dashboard/src/main/js/creation/github/GithubFlowManager.js:
```diff
@@ -114,6 +83,14 @@ export default class GithubFlowManager extends FlowManager {
+        return translate(key, opts);
+    }
+
+    getState() {
```
Is this actually used?

vivek replied:
yeah from BbRenameStep.

cliffmeyers commented on blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubPipelineCreateRequest.java:
```diff
@@ -1,204 +1,113 @@
+            errors.add(new ErrorMessage.Error("scmConfig.credentialId",
+                    ErrorMessage.Error.ErrorCodes.NOT_FOUND.toString(),
+                    "No Credentials instance found for credentialId: " + credentialId));
+        }else if(StringUtils.isBlank(scmConfig.getUri())){
```
This else if statement seems odd since the condition is checked above on line 78 and throws an exception if true?

vivek replied:
Thats bug, should have been isNotBlank, fixing.

cliffmeyers commented on blueocean-github-pipeline/src/test/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubOrgFolderTest.java:
```diff
@@ -1,457 +0,0 @@
- * @author Vivek Pandey
- */
-@RunWith(Parameterized.class)
-public class GithubOrgFolderTest extends GithubMockBase {
```
Do I read the diff correctly that these tests were just migrated over to `GithubPipelineCreateRequestTest`?

vivek replied:
yes

cliffmeyers commented on blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/OrganizationFolderRunContainerImpl.java:
```diff
@@ -33,7 +38,17 @@ public Link getLink() {
+                queueItem,
+                pipeline,
+                1
+        ).toRun();
```
Could you explain this change? I figured this class might just be removed along with other org folder stuff?

vivek replied:
Users who have pre-existing blueocean created org folder pipeline, we want them to be able to be able to edit and trigger indexing. This change triggers org folder indexing.

cliffmeyers replied:
Ok cool, so re-indexing of Org Folder can be done by invoking the "runs" API of the Org Folder? That is a huge improvement over calling the creation API to do it! 👏

vivek replied:
yes:)

cliffmeyers commented on pom.xml:
```diff
@@ -575,7 +575,9 @@
-          <version>0.3.0-beta-2</version>
+          <!-- TODO: remove this pipeline editor snapshot build with released
+version -->
+          <version>0.3.0-beta-3-20170818.000955-1</version>
```
This probably needs a new snapshot based off 0.3.0-beta-4

vivek replied:
right, fixing.

vivek commented:
@cliffmeyers I have fixed the issues pointed by you. Also modified acceptance tests for GitHub/ghe pipelines to factor in org folder to mbp change. Manual test of github/ghe look good, lets see if we get green on ATH.

cliffmeyers commented:
@vivek FYI there seem to be Jacoco failures in blueocean-github-pipeline

vivek commented:
@cliffmeyers adding more tests.

vivek commented:
@michaelneale issue reported by you is fixed, mind giving it a try?

michaelneale commented:
ok, seem better now @vivek 
One :bug: is that when I go to the configure page with classic, it shows up not as github but as api.github.com - which seems not quite right (it the type should be GitHub in the drop down). 

I am also investigating webhooks (which are a problem with org folders) as well, so may be able to fix that here. 

Finally the scope of the requested token from github does not include ability to setup webhook... (not necessarily related to this, btu thought I would look at all of this at once).

cliffmeyers commented:
I believe that api.github.com is showing in classic because that value is sent via the apiUrl parameter. Previously the UI would omit that parameter and it would default to GitHub.

michaelneale commented:
@cliffmeyers is there any harm in that?

vivek commented:
@michaelneale, @cliffmeyers yeah, side effect of requiring apiUrl. With org folder we had client send null for GitHub. We are not changing this requirement as it makes code predictable and not special case null for GitHub. 

Here is the code that creates endpoint during pipeline creation:

```java
config.setEndpoints(ImmutableList.of(new Endpoint(apiUrl, apiUrl)));
```
To keep the old behavior, backend can skip adding Endpoint if its GitHub apiUrl, in that case branch source will return `GitHub` in the drop down, as implemented in branch source, see https://github.com/vivek/github-branch-source-plugin/blob/3256676fa6a47d342a0442c59555df9eecde0b68/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java#L1457.

vivek commented:
@michaelneale I need to check if there is something extra I need to setup in branch source to enable web hook. We need extra scope for repo web hook, perhaps `write:repo_hook` as described at: https://developer.github.com/changes/2014-02-10-repo-hook-scopes/. Will look in to it tomorrow.

i386 commented:
> To keep the old behavior, backend can skip adding Endpoint if its GitHub apiUrl

That seems like a reasonable approach 👍

michaelneale commented:
@vivek yes - I am not sure if the GitHub in the drop down matters or not, but more an observation. 
I am still testing out webhooks (with tylers help)

the requested auth scope is easy to fix: https://issues.jenkins-ci.org/browse/JENKINS-46365
but there seems to be also some underlying problem with setting up webhooks with org folder (not blue ocean): https://issues.jenkins-ci.org/browse/JENKINS-46366

Given we are moving off org folders I am focussing on MBP - to make sure it is right.

i386 commented:
> Given we are moving off org folders I am focussing on MBP - to make sure it is right.

If it is broken in MBP then we should schedule an immediate fix for that.

michaelneale commented:
@i386 any idea why the build is suddenly failing with that jacoco error?

michaelneale commented:
ok I have reproduced the webhook thing, outside of blue ocean. tracking that separately - so not part of this issue. The permission scope for asking for webhook permissions probably should be included though... 

So excited this is looking great (except for the jacoco failures)

cliffmeyers commented:
Agreed regarding `apiUrl`. I think the Jacoco stuff is failing since a number of unit tests were deleted and we might be missing coverage on a few classes now?

vivek commented:
@michaelneale We can add check for web hook permission during Github/GHE credential add step. Do we know which of the scopes is sufficient to setup web hook: `admin:repo_hook` or `read:repo_hook` or `write:repo_hook` by GitHub branch source? If we are waiting for it to be fixed in upstream, we can also add this scope check in separate PR?

vivek commented:
@michaelneale pretty frustrating ATH failures in past few builds. Locally they pass. Can you run locally and see if it fails for you?

michaelneale commented:
@vivek yes running one at a time they work - I wlil run it as a whole and check... it seems worse on this branch than others though, would you say?

vivek commented:
@michaelneale thats right failing on this branch more often but thats random too. I ran complete test suites and they passed - except private repo ones. The changes in this PR do not even relate to failure in smoke test or integration test. So yeah trying to figure why its failing.

vivek commented:
@michaelneale ugh, again smoke and integration tests failed. https://ci.blueocean.io/blue/organizations/jenkins/blueocean/detail/JENKINS-45240_remove_github-org/16/tests. 

There were couple of changes in ATH on this branch but very local to GitHub creation tests: https://github.com/jenkinsci/blueocean-plugin/pull/1336/files#diff-9ad3f3e923bad131faeb0499f2906efb.

vivek commented:
the two failure in this run is more of single failure, where it fails waiting to find  `j_username` element.

```
master49666|Aug 23, 2017 11:33:57 PM org.jenkinsci.plugins.ssegateway.sse.EventDispatcherFactory start WARNING: No CrumbIssuer on Jenkins instance. Some POSTs might not work.
23:33:58 INFO  -  [0;31m???[0m Timed out while waiting for element <#j_username> to be present for 20000 milliseconds. [1;37m - expected [0;32m"visible"[0m[0m but got: [0;31m"not found"[0m
23:33:58 INFO  - [0;90m    at Object.exports.command (/home/ubuntu/slave/workspace/INS-45240_remove_github-org-A7T73VK2AVYSNVXCQPQZMKJRY34SY6DKOTEJOSCBU2DB7S3KUJCQ/acceptance-tests/src/main/js/custom_commands/login.js:3:10)[0m
23:33:58 INFO  - 
23:33:58 INFO  - 
23:33:58 INFO  - [0;31mFAILED: [0m [0;31m1[0m assertions failed (22.805s)
23:33:58 INFO  - 
23:33:58 INFO  - [1;31m _________________________________________________[0m
23:33:58 INFO  - 
23:33:58 INFO  -  [1;31mTEST FAILURE:[0m  [0;31m1[0m assertions failed, [0;32m965[0m passed. (18m 31s)
23:33:58 INFO  - 
23:33:58 INFO  - [0;31m ??? smoke[0m
23:33:58 INFO  - 
23:33:58 INFO  -    - Step 01 [1;33m(22.805s)[0m
23:33:58 INFO  - [0;31m   Timed out while waiting for element <#j_username> to be present for 20000 milliseconds. [1;37m - expected [0;32m"visible"[0m[0m but got: [0;31m"not found"[0m[0m
23:33:58 INFO  - [0;90m       at Object.exports.command (/home/ubuntu/slave/workspace/INS-45240_remove_github-org-A7T73VK2AVYSNVXCQPQZMKJRY34SY6DKOTEJOSCBU2DB7S3KUJCQ/acceptance-tests/src/main/js/custom_commands/login.js:3:10)[0m
23:33:58 INFO  - [0;36m   SKIPPED:[0m
23:33:58 INFO  -    - Step 03
23:33:58 INFO  -    - Step 04
23:33:58 INFO  -    - Step 05
23:33:58 INFO  - 
23:33:58 INFO  - Disconnected from the Jenkins SSE Gateway.
23:33:58 INFO  - >> Test suite "done". Okay to start next test.
23:33:58 INFO  - [23:33:58] 'default' errored after 19 min
23:33:58 INFO  - [23:33:58] Error: Command `nightwatch --retries 5 --suiteRetries 2` failed with exit code 1
23:33:58 INFO  -     at ChildProcess.exithandler (child_process.js:206:12)
23:33:58 INFO  -     at emitTwo (events.js:106:13)
23:33:58 INFO  -     at ChildProcess.emit (events.js:191:7)
23:33:58 INFO  -     at maybeClose (internal/child_process.js:852:16)
23:33:58 INFO  -     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:5)
**** NightwatchRunner - failure.
-------------- NightwatchRunner <<End>> --------------
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1,112.383 s <<< FAILURE! - in io.blueocean.ath.nightwatch.NightwatchTest
[ERROR] runIntegrationTest(io.blueocean.ath.nightwatch.NightwatchTest)  Time elapsed: 1,112.383 s  <<< ERROR!
com.github.eirslett.maven.plugins.frontend.lib.TaskRunnerException: 'gulp.js ' failed. (error code 1)
	at io.blueocean.ath.nightwatch.NightwatchTest.run(NightwatchTest.java:66)
	at io.blueocean.ath.nightwatch.NightwatchTest.runIntegrationTest(NightwatchTest.java:78)
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   NightwatchTest.runIntegrationTest:78->run:66 ? TaskRunner 'gulp.js ' failed. (...
[INFO] 
[ERROR] Tests run: 15, Failures: 0, Errors: 1, Skipped: 1
```

michaelneale commented:
@vivek yes it seems at least more consistent with that area now... I am rebuilding and will run full thing locally to see if I can catch it in the act. It may just be "bad luck". 

One thing the nightwatch does is that it does retries. Sometimes those retries are never going to work, so perhaps there is some other instability which cases the retry to happen, but the smoke test isn't retry friendly... I know when I run it individually it is fine.

vivek commented:
@michaelneale You are right, I can see that very first run failed at:

```
23:27:34 INFO  -  [0;31m???[0m Timed out while waiting for element <.pipelines-table .JTable-row[data-pipeline='my-pipeline']> to be present for 20000 milliseconds. [1;37m - expected [0;32m"visible"[0m[0m but got: [0;31m"not found"[0m
23:27:34 INFO  - [0;90m    at Object.module.exports.commands.assertJob (/home/ubuntu/slave/workspace/INS-45240_remove_github-org-A7T73VK2AVYSNVXCQPQZMKJRY34SY6DKOTEJOSCBU2DB7S3KUJCQ/acceptance-tests/src/main/js/page_objects/blueocean/bluePipelines.js:42:14)
23:27:34 INFO  -     at Object.module.exports.Step 02 (/home/ubuntu/slave/workspace/INS-45240_remove_github-org-A7T73VK2AVYSNVXCQPQZMKJRY34SY6DKOTEJOSCBU2DB7S3KUJCQ/acceptance-tests/src/test/js/smoke.js:37:27)[0m
```
and then retries started reporting different failures. Above is basically selecting pipeline from the pipelines table.

vivek commented:
Looks like this run failed at the same place:(

![image](https://user-images.githubusercontent.com/38139/29644327-1970ccd6-882a-11e7-96cd-c752a3b5b37a.png)

cliffmeyers commented:
Is it feasible to rewrite this specific test in Java Selenium?

michaelneale commented:
hrm nope - I am trying another branch where I put all the steps rolled into one - to see if that helps. 

Given what smoke.js is covering, I think we can ditch it, agree @imeredith @vivek ?

vivek commented:
@michaelneale @imeredith I am ok removing smoke.js test.  Most part of it are testing classic UI and blueocean side of things are anyways tested at other places.

michaelneale commented on acceptance-tests/src/main/java/io/blueocean/ath/pages/blue/GithubCreationPage.java:
```diff
@@ -130,10 +130,7 @@ public void completeCreationFlow(String apiKey, String org, String pipeline, boo
         }
         selectOrganization(org);
 
-        wait.until(singlePipelineBtn,120000)
```
there is no wait now - and there is an intermittent failure on create so this may actually have made the test less stable (it might be trying to do something before it gets a chance to load the screen)

vivek replied:
Well earlier it was single vs auto discovery and they needed organization to be selected. Now, there is already delay with selecting repo from drop down so I thought it was alright removing it. I will look at it tomorrow and if it helps will add delay.

michaelneale replied:
ah right, yes.

michaelneale commented:
Am adding in a retry that could have been there... getting closer.
